daemon/images.go

```diff
@@ -129,28 +128,34 @@ func (daemon *Daemon) Images(imageFilters filters.Args, all bool, withExtraAttrs
 		}
 
 		newImage := newImage(img, size)
+		didMatch := false
 
 		for _, ref := range daemon.referenceStore.References(id.Digest()) {
-			if imageFilters.Include("reference") {
+			if _, ok := ref.(reference.Canonical); ok {
+				newImage.RepoDigests = append(newImage.RepoDigests, ref.String())
+			}
+			if _, ok := ref.(reference.NamedTagged); ok {
+				newImage.RepoTags = append(newImage.RepoTags, ref.String())
+			}
+			if !didMatch && imageFilters.Include("reference") {
 				var found bool
 				var matchErr error
 				for _, pattern := range imageFilters.Get("reference") {
```

Comment:
I was actually looking at this; Isn't out behavior that if the _same_ filter is provided twice, that all provided filters should match? i.e. `--filter reference=foo:latest --filter reference=bar:latest` should only match images that have _both_ `foo:latest` and `bar:latest`?

Compared this with labels;

```
$ docker volume create --label one --label two foo
foo

$ docker volume create --label one bar
bar

$ docker volume ls --filter label=one --filter label=two
DRIVER              VOLUME NAME
local               foo
```